airbyte-integrations/connectors/destination-jdbc/src/main/java/io/airbyte/integrations/destination/jdbc/copy/s3/S3Copier.java

```diff
@@ -0,0 +1,210 @@
+  // WARNING: Too large a part size can cause potential OOM errors.
+  private static final int PART_SIZE_MB = 10;
+
+  private final String s3StagingFile;
+  private final DestinationSyncMode destSyncMode;
+  private final String schemaName;
+  private final String streamName;
+  private final AmazonS3 s3Client;
+  private final JdbcDatabase db;
+  private final S3Config s3Config;
+  private final ExtendedNameTransformer nameTransformer;
+  private final SqlOperations sqlOperations;
+  private final StreamTransferManager multipartUploadManager;
+  private final MultiPartOutputStream outputStream;
+  private final CSVPrinter csvPrinter;
+  private final String tmpTableName;
+
+  public S3Copier(String stagingFolder,
+                  DestinationSyncMode destSyncMode,
+                  String schema,
+                  String streamName,
+                  AmazonS3 client,
+                  JdbcDatabase db,
+                  S3Config s3Config,
+                  ExtendedNameTransformer nameTransformer,
+                  SqlOperations sqlOperations)
+      throws IOException {
+    this.destSyncMode = destSyncMode;
+    this.schemaName = schema;
+    this.streamName = streamName;
+    this.s3Client = client;
+    this.db = db;
+    this.s3Config = s3Config;
+    this.nameTransformer = nameTransformer;
+    this.sqlOperations = sqlOperations;
+
+    this.s3StagingFile = String.join("/", stagingFolder, schemaName, streamName);
+
+    this.tmpTableName = nameTransformer.getTmpTableName(streamName);
+    // The stream transfer manager lets us greedily stream into S3. The native AWS SDK does not
+    // have support for streaming multipart uploads;
+    // The alternative is first writing the entire output to disk before loading into S3. This is not
+    // feasible with large tables.
+    // Data is chunked into parts. A part is sent off to a queue to be uploaded once it has reached it's
+    // configured part size.
+    // Memory consumption is queue capacity * part size = 10 * 10 = 100 MB at current configurations.
+    this.multipartUploadManager =
+        new StreamTransferManager(s3Config.getBucketName(), s3StagingFile, client)
+            .numUploadThreads(DEFAULT_UPLOAD_THREADS)
+            .queueCapacity(DEFAULT_QUEUE_CAPACITY)
+            .partSize(PART_SIZE_MB);
+    // We only need one output stream as we only have one input stream. This is reasonably performant.
+    // See the above comment.
+    this.outputStream = multipartUploadManager.getMultiPartOutputStreams().get(0);
+
+    var writer = new PrintWriter(outputStream, true, StandardCharsets.UTF_8);
+    this.csvPrinter = new CSVPrinter(writer, CSVFormat.DEFAULT);
+  }
+
+  @Override
+  public void write(UUID id, String jsonDataString, Timestamp emittedAt) throws Exception {
+    csvPrinter.printRecord(id, jsonDataString, emittedAt);
+  }
+
+  @Override
+  public String copyToTmpTableAndPrepMergeToFinalTable(boolean hasFailed) throws Exception {
```

Comment:
oh so this does handle copying into the final table? `PrepMerge` makes me think it preps something but doesn't actually put data in the final table. looking at the impl it seems like it actually does the merge.

Reply:
there's a pretty detailed lifecycle in here. could we do more to make it more approachable? i feel like i'm working pretty hard to understand what each step is supposed to be. having a list of each step in the lifecycle would be helfpul.